engine/runtime/src/main/scala/org/enso/compiler/pass/lint/NoSelfInStatic.scala

```diff
@@ -0,0 +1,104 @@
+package org.enso.compiler.pass.lint
+
+import org.enso.compiler.context.{InlineContext, ModuleContext}
+import org.enso.compiler.core.{CompilerError, IR}
+import org.enso.compiler.pass.IRPass
+import org.enso.compiler.pass.desugar.GenerateMethodBodies
+
+/** This linting pass ensures that `self` argument is not used in static methods.
+  *
+  * This pass requires the context to provide:
+  * - Nothing
+  */
+object NoSelfInStatic extends IRPass {
+  override type Metadata = IRPass.Metadata.Empty
+  override type Config   = IRPass.Configuration.Default
+
+  override val precursorPasses: Seq[IRPass] =
+    Seq(GenerateMethodBodies)
+
+  override val invalidatedPasses: Seq[IRPass] = List()
+
+  override def runModule(
+    ir: IR.Module,
+    moduleContext: ModuleContext
+  ): IR.Module = {
+    ir.copy(
+      bindings = ir.bindings.map {
+        case method: IR.Module.Scope.Definition.Method.Explicit
+            if isStaticMethod(method) =>
+          method.copy(
+            body = method.body.transformExpressions(transformSelfToError)
+          )
+        case method: IR.Module.Scope.Definition.Method.Binding =>
+          throw new CompilerError("Unreachable: Method.Binding should not be present in this pass: " + method)
```

Comment:
I would keep the error message simpler:
```suggestion
          throw new CompilerError(s"unexpected Method.Binding $method present in pass NoSelfInStatic")
```